src/compiler/checker.ts

```diff
@@ -7205,6 +7434,57 @@ module ts {
             return resolveCall(node, callSignatures, candidatesOutArray);
         }
 
+        function resolveDecorator(node: Decorator, candidatesOutArray: Signature[]): Signature {
+            let funcType = checkExpression(node.expression);
+            let apparentType = getApparentType(funcType);
+            if (apparentType === unknownType) {
+                return resolveErrorCall(node);
+            }
+            
+            let callSignatures = getSignaturesOfType(apparentType, SignatureKind.Call);
+            if (funcType === anyType || (!callSignatures.length && !(funcType.flags & TypeFlags.Union) && isTypeAssignableTo(funcType, globalFunctionType))) {

```

Comment:
Generally I'm using the same approach that resolveTaggedTemplateExpression uses, which avoids some of the work that resolveCallExpression does.